### ISSUE TITLE
Add more thorough python binding Matrix tests

### ISSUE THREAD

**Issue by moxcodes** [sxs-collaboration/spectre#2162]
# Feature request:

Currently, the python binding H5 interface is more demanding of certain features of the python Matix bindings than the Matrix test (the H5 interface cares about data layout, for instance), so subtle interface bugs end up breaking the H5 test instead. It would be more helpful if that type of error showed up closer to the source.

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

**Comment by nilsvu** [comment]
@moxcodes could you be a bit more specific, perhaps give an example of an issue you encountered?

**Comment by moxcodes** [comment]
In PR #2114, I encountered a trouble when the underlying data layout changed (the spacing between contiguous segments became `.spacing()` rather than `.rows()` in the underlying blaze -- an unintentional defect in blaze, but that's another story), the python Matrix tests were unaffected, presumably because all of the element accesses that they do don't actually care about the underlying data layout. However, the H5 bindings did care about the layout, and they failed with mysterious data gaps.
It's not clear to me how generalizable the trouble is -- it might be that this is a narrow corner case and would require only a small test change, but because there's some dependency on the data layout of Matrix in other python bindings, it would be nice for the data structure test to cover that interface feature as well.
Another example to check is that if the ordering of `Matrix` is changed to `rowMajor` in the c++, I'd like the python checks to fail (I haven't tried that one - that request might already be satisfied).

**Comment by nilsvu** [comment]
I'm not saying the Python tests are sufficiently complete (they probably aren't; we have so far mostly added binding tests to have at least _some_, without caring too much about completeness), but what you're saying sounds to me like the tests that should be more thorough are the C++ tests. The Pybindings tests should only check the bindings, i.e. that the Python interface is the same as the C++ interface and that interactions with other Python code (e.g. NumPy) works reasonably well. The Pybindings tests should _not_ test the C++ code _again_, that's what the C++ tests are for. So I consider it correct behavior when a C++ test fails, but the Pybindings test succeeds - indeed that's good, because it means the issue is in the C++ code, not in the bindings code. Do you agree here, and does this actually apply to your issue?

**Comment by moxcodes** [comment]
The error here was in `DataStructures/Python/Matrix.cpp` not having been adjusted for the underlying C++ change. For everything else, I agree that the python tests are not responsible for catching errors in the C++ code.
However, here the case was that the full set of C++ libraries were in a consistent and correct state, all C++ tests passing (and correctly), but the interface exposed to the python bindings did not account for the layout change. The python tests of the data structures, then, were not dependent on the layout, but the python h5 tests were dependent on the layout, so failed.
What I would have hoped for is that the python test for the Matrix fail when the interface it provides to the python binding for H5 is in an inconsistent state.
Does that make sense?

**Comment by nilsvu** [comment]
Yes that makes a lot of sense, thanks for detailing! Which part of `DataStructures/Python/Matrix.cpp` needed adjusting?

**Comment by moxcodes** [comment]
The part that caused trouble in the translation was the exposure of the python buffer:
```
...
      .def_buffer([](Matrix& matrix) {
        return py::buffer_info(
            matrix.data(),
            // Size of one scalar
            sizeof(double), py::format_descriptor<double>::format(),
            // Number of dimensions
            2,
            // Size of the buffer
            {matrix.rows(), matrix.columns()},
            // Stride for each index (in bytes). Data is stored
            // in column-major layout (see `Matrix.hpp`).
            {sizeof(double), sizeof(double) * matrix.rows()});
      })
...
```
The fix in this case was changing `matrix.rows()` to `matrix.columns()` in the last line; but I anticipate that this will have trouble also if the underlying matrix class were to change storage order to `blaze::rowMajor`.
